### PR TITLE
Clarify `marginal_counts` documentation and fix default behavior (#6230)

### LINKED ISSUE #6230: Inconsistent ordering of marginal counts
### Information

- **Qiskit Terra version**: 0.24.1
- **Python version**:
- **Operating system**:

### What is the current behavior?
`marginal_counts` ordering does not obey a standard convention when returning its output.  For example, consider the counts:

```
counts = {'10011':1, '01100':1}
```
using `indices = [0,1,2,3,4]` give the identity mapping, suggesting that the 0th index in the indices list maps the input counts bit to the least significant bit in the new marginal counts, the next least significant is set by the 1st index value etc.  However, this does not hold true in general.  Namely `indices = [0,1,2,4,3]` does not follow the prescribed convention and returns:

```
{'10011': 1, '01100': 1}
```

whereas one would expect the following:

```
{'01011': 1, '10100': 1}
```
i.e. the output bits are in the order q3q4q2q1q0

I believe it is due to the line: https://github.com/Qiskit/qiskit-terra/blob/40ffd2b07c03e220a174b2bfd0bf8c23aa021c5f/qiskit/result/utils.py#L111

that implicitly reorders any permutation of indices.  This is mentioned no where in the docs, and is confusing  

### What is the expected behavior?
I would expect the order of indices given to have a one to one mapping to those in the output.

### PR DESCRIPTION
**Pull request by mattwright99**
### Summary

Fixes #6230 

### Details and comments

The original issue was to clarify documentation of the `marginal_counts` function (which was done) but I discovered another issue when calling the function with `indices=None` (the default). The issue arose from calling `len(indices)` and was fixed by checking if `indices` is `None` at the beginning and then returning the original result object because the function would have no affect.

### CODE REVIEW

File: qiskit/result/utils.py [hunk 1]
```diff
@@ -26,8 +26,9 @@ def marginal_counts(result, indices=None, inplace=False, format_marginal=False):
     Args:
         result (dict or Result): result to be marginalized
             (a Result object or a dict(str, int) of counts).
-        indices (list(int) or None): The bit positions of interest
-            to marginalize over. If ``None`` (default), do not marginalize at all.
+        indices (set(int) or None): The bit positions of interest. Bit indices not
+            spcified will be ignored. If ``None`` (default), do not marginalize at all,
```
Comment (javabster) [inline qiskit/result/utils.py:30]:
```suggestion
            specified will be ignored. If ``None`` (default), do not marginalize at all,
```

**Review by javabster — COMMENTED** [review]
Thanks for working on this @mattwright99! Can you add a test please?

**Comment by mattwright99** [comment]
@javabster added some tests! The function is also tested pretty extensivelty in the [`Counts` object unit tests](https://github.com/Qiskit/qiskit-terra/blob/main/test/python/result/test_counts.py) so I just added a few basic tests to make sure the `marginal_counts` function operates as it should.

File: test/python/result/test_marginal_counts.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+"""Test marginal_counts function."""
+
+import unittest
+
+from qiskit.result.utils import marginal_counts
+
+
+class TestMarginalCounts(unittest.TestCase):
+    def test_default(self):
```
Comment (javabster) [inline test/python/result/test_marginal_counts.py:23]: Thanks for the tests @mattwright99, but I actually meant could you add a test for the case that threw the error previously (I believe it was the `indices=None` case?). Sorry if I wasn't clear before!

Also it looks like CI is failing because of linting issues. You can run `tox -eblack` to fix

File: test/python/result/test_marginal_counts.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# pylint: disable=missing-class-docstring,missing-function-docstring
+
+"""Test marginal_counts function."""
+
+import unittest
+
+from qiskit.result.utils import marginal_counts
+
+
+class TestMarginalCounts(unittest.TestCase):
```
Comment (jakelishman) [inline test/python/result/test_marginal_counts.py:22]:
```suggestion
from qiskit.result.utils import marginal_counts
from qiskit.test import QiskitTestCase


class TestMarginalCounts(QiskitTestCase):
```

File: test/python/result/test_marginal_counts.py (hunk 2 above)
Comment (mattwright99) [reply test/python/result/test_marginal_counts.py:22]: Oh I see, sorry. Since `indices=None` is the default setting, it is tested in the very first function `test_default`. Would you like me to be more explicit?

Comment (javabster) [reply test/python/result/test_marginal_counts.py:22]: When I run your default test against main I don't get any error, so it doesn't seem to be covering the edge case of the original error. I think it only crops up when you pass a result object, not just the raw counts

Comment (mattwright99) [reply test/python/result/test_marginal_counts.py:22]: Yes you are completely right, thanks for catching that. I will add a proper test now.

**Comment by taalexander** [comment]
@jakelishman and @javabster bumping for re-review.

File: qiskit/result/utils.py [hunk 4]
```diff
@@ -43,6 +44,9 @@ def marginal_counts(result, indices=None, inplace=False, format_marginal=False):
     Raises:
         QiskitError: in case of invalid indices to marginalize over.
     """
+    if indices is None:
+        return result
+
```
Comment (jakelishman) [inline qiskit/result/utils.py:49]: The documentation says that setting `indices=None` is equivalent to providing the set of all indices, but this isn't the case with this change; the renormalisation of the `key` strings inside `_marginalize` will happen in the latter case but not the former, but only if `result` is of type `Result`.

Given that `_marginalize` and the other branch of `marginal_counts` are correctly set up to handle `indices=None`, I think it would be better to fix the broken handling inside `marginal_counts`, so we can ensure that the outputs really are identical.

File: test/python/result/test_marginal_counts.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# This code is part of Qiskit.
```
Comment (jakelishman) [inline test/python/result/test_marginal_counts.py:1]: There are already a lot of `marginal_counts` tests in `test/python/result/test_result.py` - can you put these tests with them in that format to keep everything together?

File: test/python/result/test_marginal_counts.py (hunk 2 above)
Comment (jakelishman) [reply test/python/result/test_marginal_counts.py:33]: Please could this test have a more descriptive name of what's being tested?  It's actually a test that passing `indices=None` succeeds - the fact that that's the default is not really important.

**Review by jakelishman — CHANGES_REQUESTED** [review]
Thanks for looking at this, and sorry for the delay.  In the future, feel free to tag us to re-request review once you're done adding commits.

**Comment by javabster** [comment]
Hi @mattwright99 how is this going? Let us know if you need any further assistance 😄

**Comment by 1ucian0** [comment]
Since the PR looks stalled, I'm closing it. @mattwright99 if you still are working on this, let us know and it can be reopened!
